// hadron-document/lib/editor/index.js
"use strict";

var StandardEditor = require('./standard');

var StringEditor = require('./string');

var Decimal128Editor = require('./decimal128');

var Int32Editor = require('./int32');

var Int64Editor = require('./int64');

var DoubleEditor = require('./double');

var DateEditor = require('./date');

var NullEditor = require('./null');

var UndefinedEditor = require('./undefined');

var ObjectIdEditor = require('./objectid');

var init = function init(element) {
  return {
    Standard: new StandardEditor(element),
    String: new StringEditor(element),
    Decimal128: new Decimal128Editor(element),
    Date: new DateEditor(element),
    Double: new DoubleEditor(element),
    Int32: new Int32Editor(element),
    Int64: new Int64Editor(element),
    Null: new NullEditor(element),
    Undefined: new UndefinedEditor(element),
    ObjectId: new ObjectIdEditor(element)
  };
};

Object.assign(init, {
  DateEditor: DateEditor,
  StandardEditor: StandardEditor,
  StringEditor: StringEditor,
  Decimal128Editor: Decimal128Editor,
  DoubleEditor: DoubleEditor,
  Int32Editor: Int32Editor,
  Int64Editor: Int64Editor,
  NullEditor: NullEditor,
  UndefinedEditor: UndefinedEditor,
  ObjectIdEditor: ObjectIdEditor
});
module.exports = init;
