// @mongodb-js/compass-components/lib/components/cancel-loader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const emotion_1 = require("@leafygreen-ui/emotion");
const react_1 = __importDefault(require("react"));
const cancelLoaderStyle = (0, emotion_1.css)({
    height: '100%',
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
    padding: '10px 0',
});
const progressTextStyle = (0, emotion_1.css)({
    margin: '0 auto 30px',
    textAlign: 'left',
    fontSize: '24px',
    color: '#807f7f',
    fontWeight: 200,
});
const spinnerStyle = (0, emotion_1.css)({
    animation: 'fa-spin 2s infinite linear',
    width: '40px',
    textAlign: 'center',
    display: 'inline-block',
    textRendering: 'auto',
    font: 'normal normal normal 20px/1 FontAwesome',
    '&::before': {
        boxSizing: 'border-box',
        content: '"\\f1ce"',
    },
});
const buttonStyle = (0, emotion_1.css)({
    padding: '0 10px 0 10px',
    height: '28px',
    fontWeight: 'bold',
    fontSize: '13px',
    lineHeight: '26px',
    textTransform: 'uppercase',
    backgroundColor: 'transparent',
    border: '1px solid #13AA52',
    borderRadius: '3px',
    boxShadow: 'none',
    color: '#13AA52',
    fontFamily: 'Akzidenz',
    WebkitAppearance: 'button',
    cursor: 'pointer',
    overflow: 'visible',
    margin: 0,
    boxSizing: 'border-box',
    outline: 'none',
});
function CancelLoader({ dataTestId, progressText, cancelText, onCancel, }) {
    return (react_1.default.createElement("div", { "data-testid": dataTestId, className: cancelLoaderStyle },
        react_1.default.createElement("div", { className: progressTextStyle },
            react_1.default.createElement("i", { className: spinnerStyle }),
            progressText),
        react_1.default.createElement("div", null,
            react_1.default.createElement("button", { className: buttonStyle, onClick: onCancel }, cancelText))));
}
exports.default = CancelLoader;
