// @mongodb-js/compass-components/lib/components/error-boundary.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorBoundary = void 0;
const react_1 = __importDefault(require("react"));
const emotion_1 = require("@leafygreen-ui/emotion");
const tokens_1 = require("@leafygreen-ui/tokens");
const leafygreen_1 = require("./leafygreen");
const errorContainerStyles = (0, emotion_1.css)({
    padding: tokens_1.spacing[3],
    width: '100%',
});
class ErrorBoundary extends react_1.default.Component {
    constructor() {
        super(...arguments);
        this.state = {
            error: null,
        };
    }
    static getDerivedStateFromError(error) {
        return {
            error,
        };
    }
    componentDidCatch(error, errorInfo) {
        if (this.props.onError) {
            this.props.onError(error, errorInfo);
        }
    }
    render() {
        const { displayName } = this.props;
        const { error } = this.state;
        if (error) {
            return (react_1.default.createElement("div", { className: errorContainerStyles },
                react_1.default.createElement(leafygreen_1.Banner, { variant: "danger" },
                    "An error occurred while rendering",
                    displayName ? ` ${displayName}` : '',
                    ": ",
                    error.message)));
        }
        return this.props.children;
    }
}
const _ErrorBoundary = (props) => {
    var _a;
    return ((_a = process === null || process === void 0 ? void 0 : process.env) === null || _a === void 0 ? void 0 : _a.COMPASS_DISABLE_ERROR_BOUNDARY) === 'true' ? (react_1.default.createElement(react_1.default.Fragment, null, props.children)) : (react_1.default.createElement(ErrorBoundary, { ...props }));
};
exports.ErrorBoundary = _ErrorBoundary;
