// @mongodb-js/compass-components/lib/components/inline-definition.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InlineDefinition = void 0;
const react_1 = __importDefault(require("react"));
const emotion_1 = require("@leafygreen-ui/emotion");
const leafygreen_1 = require("./leafygreen");
const tooltip_1 = require("./tooltip");
const merge_props_1 = require("../utils/merge-props");
const underline = (0, emotion_1.css)({
    backgroundRepeat: 'repeat-x',
    backgroundPosition: 'center bottom',
    backgroundSize: '3px 2px',
    backgroundImage: 'radial-gradient( circle closest-side, currentColor 75%, transparent 25%)',
});
const maxWidth = (0, emotion_1.css)({
    maxWidth: '360px',
});
const InlineDefinition = ({ children, definition, tooltipProps, ...props }) => {
    return (react_1.default.createElement(tooltip_1.Tooltip, { justify: "middle", spacing: 5, trigger: ({ children: tooltip, ...triggerProps }) => {
            const merged = (0, merge_props_1.mergeProps)({ className: underline }, triggerProps, props);
            return (react_1.default.createElement("span", { ...merged },
                children,
                tooltip));
        }, ...tooltipProps },
        react_1.default.createElement(leafygreen_1.Body, { className: maxWidth }, definition)));
};
exports.InlineDefinition = InlineDefinition;
