// @mongodb-js/compass-components/lib/components/leafygreen.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Description = exports.Link = exports.Label = exports.Overline = exports.Disclaimer = exports.InlineKeyCode = exports.InlineCode = exports.Body = exports.Subtitle = exports.H3 = exports.H2 = exports.H1 = exports.Tooltip = exports.Toggle = exports.Toast = exports.TextInput = exports.TextArea = exports.Tabs = exports.Tab = exports.Cell = exports.Row = exports.TableHeader = exports.Table = exports.OptionGroup = exports.Option = exports.Select = exports.SegmentedControlOption = exports.SegmentedControl = exports.RadioGroup = exports.Radio = exports.RadioBoxGroup = exports.RadioBox = exports.MongoDBLogo = exports.MongoDBLogoMark = exports.ModalFooter = exports.Modal = exports.MenuSeparator = exports.MenuItem = exports.Menu = exports.IconButton = exports.Icon = exports.ConfirmationModal = exports.Checkbox = exports.Card = exports.Button = exports.Banner = exports.Badge = exports.AtlasLogoMark = void 0;
const use_theme_1 = require("../hooks/use-theme");
const badge_1 = __importDefault(require("@leafygreen-ui/badge"));
Object.defineProperty(exports, "Badge", { enumerable: true, get: function () { return badge_1.default; } });
const banner_1 = __importDefault(require("@leafygreen-ui/banner"));
Object.defineProperty(exports, "Banner", { enumerable: true, get: function () { return banner_1.default; } });
const button_1 = __importDefault(require("@leafygreen-ui/button"));
const checkbox_1 = __importDefault(require("@leafygreen-ui/checkbox"));
const card_1 = __importDefault(require("@leafygreen-ui/card"));
const confirmation_modal_1 = __importDefault(require("@leafygreen-ui/confirmation-modal"));
const icon_1 = __importDefault(require("@leafygreen-ui/icon"));
Object.defineProperty(exports, "Icon", { enumerable: true, get: function () { return icon_1.default; } });
const icon_button_1 = __importDefault(require("@leafygreen-ui/icon-button"));
const logo_1 = require("@leafygreen-ui/logo");
Object.defineProperty(exports, "AtlasLogoMark", { enumerable: true, get: function () { return logo_1.AtlasLogoMark; } });
Object.defineProperty(exports, "MongoDBLogoMark", { enumerable: true, get: function () { return logo_1.MongoDBLogoMark; } });
Object.defineProperty(exports, "MongoDBLogo", { enumerable: true, get: function () { return logo_1.MongoDBLogo; } });
const menu_1 = require("@leafygreen-ui/menu");
Object.defineProperty(exports, "Menu", { enumerable: true, get: function () { return menu_1.Menu; } });
Object.defineProperty(exports, "MenuSeparator", { enumerable: true, get: function () { return menu_1.MenuSeparator; } });
Object.defineProperty(exports, "MenuItem", { enumerable: true, get: function () { return menu_1.MenuItem; } });
const modal_1 = __importStar(require("@leafygreen-ui/modal"));
const radio_box_group_1 = require("@leafygreen-ui/radio-box-group");
Object.defineProperty(exports, "RadioBox", { enumerable: true, get: function () { return radio_box_group_1.RadioBox; } });
Object.defineProperty(exports, "RadioBoxGroup", { enumerable: true, get: function () { return radio_box_group_1.RadioBoxGroup; } });
const radio_group_1 = require("@leafygreen-ui/radio-group");
Object.defineProperty(exports, "Radio", { enumerable: true, get: function () { return radio_group_1.Radio; } });
const segmented_control_1 = require("@leafygreen-ui/segmented-control");
Object.defineProperty(exports, "SegmentedControlOption", { enumerable: true, get: function () { return segmented_control_1.SegmentedControlOption; } });
const select_1 = require("@leafygreen-ui/select");
Object.defineProperty(exports, "Option", { enumerable: true, get: function () { return select_1.Option; } });
Object.defineProperty(exports, "OptionGroup", { enumerable: true, get: function () { return select_1.OptionGroup; } });
const table_1 = require("@leafygreen-ui/table");
Object.defineProperty(exports, "TableHeader", { enumerable: true, get: function () { return table_1.TableHeader; } });
Object.defineProperty(exports, "Row", { enumerable: true, get: function () { return table_1.Row; } });
Object.defineProperty(exports, "Cell", { enumerable: true, get: function () { return table_1.Cell; } });
const tabs_1 = require("@leafygreen-ui/tabs");
Object.defineProperty(exports, "Tab", { enumerable: true, get: function () { return tabs_1.Tab; } });
const text_area_1 = __importDefault(require("@leafygreen-ui/text-area"));
const text_input_1 = __importDefault(require("@leafygreen-ui/text-input"));
const toast_1 = __importDefault(require("@leafygreen-ui/toast"));
Object.defineProperty(exports, "Toast", { enumerable: true, get: function () { return toast_1.default; } });
const toggle_1 = __importDefault(require("@leafygreen-ui/toggle"));
const tooltip_1 = __importDefault(require("@leafygreen-ui/tooltip"));
const typography_1 = require("@leafygreen-ui/typography");
Object.defineProperty(exports, "H1", { enumerable: true, get: function () { return typography_1.H1; } });
Object.defineProperty(exports, "H2", { enumerable: true, get: function () { return typography_1.H2; } });
Object.defineProperty(exports, "H3", { enumerable: true, get: function () { return typography_1.H3; } });
Object.defineProperty(exports, "Subtitle", { enumerable: true, get: function () { return typography_1.Subtitle; } });
Object.defineProperty(exports, "Body", { enumerable: true, get: function () { return typography_1.Body; } });
Object.defineProperty(exports, "InlineCode", { enumerable: true, get: function () { return typography_1.InlineCode; } });
Object.defineProperty(exports, "InlineKeyCode", { enumerable: true, get: function () { return typography_1.InlineKeyCode; } });
Object.defineProperty(exports, "Disclaimer", { enumerable: true, get: function () { return typography_1.Disclaimer; } });
Object.defineProperty(exports, "Overline", { enumerable: true, get: function () { return typography_1.Overline; } });
Object.defineProperty(exports, "Link", { enumerable: true, get: function () { return typography_1.Link; } });
const Button = (0, use_theme_1.withTheme)(button_1.default);
exports.Button = Button;
const Card = (0, use_theme_1.withTheme)(card_1.default);
exports.Card = Card;
const Checkbox = (0, use_theme_1.withTheme)(checkbox_1.default);
exports.Checkbox = Checkbox;
const ConfirmationModal = (0, use_theme_1.withTheme)(confirmation_modal_1.default);
exports.ConfirmationModal = ConfirmationModal;
const IconButton = (0, use_theme_1.withTheme)(icon_button_1.default);
exports.IconButton = IconButton;
const ModalFooter = (0, use_theme_1.withTheme)(modal_1.Footer);
exports.ModalFooter = ModalFooter;
const Modal = (0, use_theme_1.withTheme)(modal_1.default);
exports.Modal = Modal;
const RadioGroup = (0, use_theme_1.withTheme)(radio_group_1.RadioGroup);
exports.RadioGroup = RadioGroup;
const SegmentedControl = (0, use_theme_1.withTheme)(segmented_control_1.SegmentedControl);
exports.SegmentedControl = SegmentedControl;
const Select = (0, use_theme_1.withTheme)(select_1.Select);
exports.Select = Select;
const Table = (0, use_theme_1.withTheme)(table_1.Table);
exports.Table = Table;
const Tabs = (0, use_theme_1.withTheme)(tabs_1.Tabs);
exports.Tabs = Tabs;
const TextArea = (0, use_theme_1.withTheme)(text_area_1.default);
exports.TextArea = TextArea;
const TextInput = (0, use_theme_1.withTheme)(text_input_1.default);
exports.TextInput = TextInput;
const Toggle = (0, use_theme_1.withTheme)(toggle_1.default);
exports.Toggle = Toggle;
const Tooltip = (0, use_theme_1.withTheme)(tooltip_1.default);
exports.Tooltip = Tooltip;
const Label = (0, use_theme_1.withTheme)(typography_1.Label);
exports.Label = Label;
const Description = (0, use_theme_1.withTheme)(typography_1.Description);
exports.Description = Description;
