// @mongodb-js/compass-components/lib/components/spin-loader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpinLoader = void 0;
const react_1 = __importDefault(require("react"));
const emotion_1 = require("@leafygreen-ui/emotion");
const palette_1 = require("@leafygreen-ui/palette");
const shellLoaderSpin = (0, emotion_1.keyframes) `
  0% { transform: rotate(0deg); }
  100% { transform: rotate(360deg); }
`;
const spinLoaderStyle = (0, emotion_1.css) `
  border: 2px solid transparent;
  border-top: 2px solid ${palette_1.uiColors.gray.dark3};
  border-radius: 50%;
  padding: 0;
  margin: 0;
  box-sizing: border-box;
  display: inline-block;

  animation: ${shellLoaderSpin} 700ms ease infinite;
`;
function SpinLoader({ size = '12px' }) {
    return (react_1.default.createElement("div", { className: spinLoaderStyle, style: {
            width: size,
            height: size,
        } }));
}
exports.SpinLoader = SpinLoader;
