// @mongodb-js/compass-components/lib/components/workspace-container.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspaceContainer = void 0;
const react_1 = __importDefault(require("react"));
const emotion_1 = require("@leafygreen-ui/emotion");
const palette_1 = require("@leafygreen-ui/palette");
const use_theme_1 = require("../hooks/use-theme");
const compass_ui_colors_1 = require("../compass-ui-colors");
const workspaceContainerStyles = (0, emotion_1.css)({
    height: '100%',
    width: '100%',
    display: 'flex',
    overflow: 'auto',
});
const lightThemeStyles = (0, emotion_1.css)({
    backgroundColor: compass_ui_colors_1.gray8,
    color: palette_1.uiColors.gray.dark2,
});
const darkThemeStyles = (0, emotion_1.css)({
    backgroundColor: palette_1.uiColors.gray.dark3,
    color: palette_1.uiColors.white,
});
function UnthemedWorkspaceContainer({ className, darkMode, children, 'data-test-id': dataTestId, }) {
    return (react_1.default.createElement("div", { className: (0, emotion_1.cx)(workspaceContainerStyles, darkMode ? darkThemeStyles : lightThemeStyles, className), "data-test-id": dataTestId }, children));
}
const WorkspaceContainer = (0, use_theme_1.withTheme)(UnthemedWorkspaceContainer);
exports.WorkspaceContainer = WorkspaceContainer;
