// @mongodb-js/compass-components/lib/hooks/use-focus-ring.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useFocusRing = exports.focusRingVisible = exports.focusRing = void 0;
const tokens_1 = require("@leafygreen-ui/tokens");
const emotion_1 = require("@leafygreen-ui/emotion");
const palette_1 = require("@leafygreen-ui/palette");
const use_focus_hover_1 = require("./use-focus-hover");
exports.focusRing = (0, emotion_1.css)({
    position: 'relative',
    outline: 'none',
    '&::after': {
        position: 'absolute',
        content: '""',
        pointerEvents: 'none',
        top: 3,
        right: 3,
        bottom: 3,
        left: 3,
        borderRadius: tokens_1.spacing[1],
        boxShadow: `0 0 0 0 ${palette_1.uiColors.focus}`,
        transition: 'box-shadow .16s ease-in',
    },
});
exports.focusRingVisible = (0, emotion_1.css)({
    '&::after': {
        boxShadow: `0 0 0 3px ${palette_1.uiColors.focus}`,
        transitionTimingFunction: 'ease-out',
    },
});
function useFocusRing() {
    const [focusProps, focusState] = (0, use_focus_hover_1.useFocusState)();
    return {
        ...focusProps,
        className: (0, emotion_1.cx)(exports.focusRing, focusState === use_focus_hover_1.FocusState.FocusVisible && exports.focusRingVisible),
    };
}
exports.useFocusRing = useFocusRing;
