// @mongodb-js/compass-components/lib/hooks/use-sort.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSortedItems = exports.useSortControls = void 0;
const react_1 = __importStar(require("react"));
const utils_1 = require("@react-aria/utils");
const __1 = require("..");
const controlsContainer = (0, __1.css)({
    display: 'flex',
    alignItems: 'center',
    gap: __1.spacing[2],
});
const label = (0, __1.css)({
    margin: '0 !important',
    padding: '0 !important',
});
const select = (0, __1.css)({
    '& > button': {
        margin: 0,
    },
});
function useSortControls(items, options) {
    var _a, _b;
    const labelId = (0, utils_1.useId)('Sort by');
    const controlId = (0, utils_1.useId)();
    const [sortState, dispatch] = (0, react_1.useReducer)((state, action) => {
        if (action.type === 'change-name' && action.name !== state.name) {
            return {
                ...state,
                name: action.name,
            };
        }
        if (action.type === 'change-order') {
            return {
                ...state,
                order: (state.order * -1),
            };
        }
        return state;
    }, { name: (_b = (_a = items[0]) === null || _a === void 0 ? void 0 : _a.name) !== null && _b !== void 0 ? _b : null, order: 1 });
    const sortControls = (0, react_1.useMemo)(() => {
        var _a;
        const glyph = sortState.order === -1
            ? 'SortDescending'
            : sortState.order === 1
                ? 'SortAscending'
                : 'Unsorted';
        const longestLabel = Math.max(...items.map((item) => item.label.length));
        return (react_1.default.createElement("div", { className: controlsContainer },
            react_1.default.createElement(__1.Label, { id: labelId, htmlFor: controlId, className: label }, "Sort by"),
            react_1.default.createElement(__1.Select, { disabled: options === null || options === void 0 ? void 0 : options.isDisabled, id: controlId, "aria-labelledby": labelId, allowDeselect: false, className: (0, __1.cx)(select, (0, __1.css)({ minWidth: `calc(${longestLabel}ch + ${__1.spacing[6]}px)` })), onChange: (value) => {
                    dispatch({ type: 'change-name', name: value || null });
                }, defaultValue: (_a = sortState.name) !== null && _a !== void 0 ? _a : undefined }, items.map((item) => (react_1.default.createElement(__1.Option, { key: item.name, value: item.name }, item.label)))),
            react_1.default.createElement(__1.Button, { "aria-label": glyph, title: glyph, rightGlyph: react_1.default.createElement(__1.Icon, { glyph: glyph }), onClick: () => {
                    dispatch({ type: 'change-order' });
                }, disabled: sortState.name === null || (options === null || options === void 0 ? void 0 : options.isDisabled) })));
    }, [sortState, items, labelId, controlId, options === null || options === void 0 ? void 0 : options.isDisabled]);
    return [
        sortControls,
        (options === null || options === void 0 ? void 0 : options.isDisabled) ? { name: null, order: 1 } : sortState,
    ];
}
exports.useSortControls = useSortControls;
function sortUnknown(a, b, order) {
    return (a === b ? 0 : a < b ? -1 : 1) * order;
}
function sortString(a, b, order) {
    return a.localeCompare(b) * order;
}
function useSortedItems(items, { name, order, }, sortFn = null) {
    return (0, react_1.useMemo)(() => {
        return [...items].sort((a, b) => {
            if (!name) {
                return 0;
            }
            if (typeof a[name] === 'undefined') {
                return 1;
            }
            const sort = sortFn && sortFn[name];
            if (sort) {
                return sort(a[name], b[name], order);
            }
            if (typeof a[name] === 'string') {
                return sortString(a[name], b[name], order);
            }
            return sortUnknown(a[name], b[name], order);
        });
    }, [items, name, sortFn, order]);
}
exports.useSortedItems = useSortedItems;
